Replace Mbed PSA with TF-M PSA

Patater opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Replace the Mbed implementation of PSA with the implementation from TrustedFirmware-M (TF-M).

PSA is and continues to be provided as an experimental feature (under FEATURE_EXPERIMENTAL_API). The PSA implementation provided by TF-M is not guaranteed to provide the same functionality or API as the previous Mbed PSA implmementation being replaced.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
v8-M NS targets are now the only v8-M targets supported in Mbed OS for PSA platforms, as the TF-M build system outputs the S binary that Mbed OS consumes as-is. As such, targets like `ARM_MUSCA_A1_NS` have been renamed to `ARM_MUSCA_A1` (although old names are temporarily provided for backwards compatibility).

#### Migration actions required <!-- Optional -->
How targets enable PSA has changed. Please refer to features/FEATURE_PSA/supporting_psa_in_mbed-os.md for updated instructions on adding PSA to your target as a new target porting person.

### Documentation <!-- Required -->

Porting guide updates for PSA at features/FEATURE_PSA/supporting_psa_in_mbed-os.md

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@devran01 @jainvikas8 

----------------------------------------------------------------------------------------------------------------

urutva committed:
psa: Replace Mbed PSA with TF-M

Add TF-M to Mbed OS, replacing the previous PSA implementation for
TF-M-capable targets. This commit adds files imported from TF-M, without
modification. The version of TF-M imported can be found in
`features/FEATURE_PSA/TARGET_TFM/VERSION.txt`.

These changes switch to TF-M as the sole PSA implementation for v8-M and
dual core targets, with TF-M running on the secure side and Mbed OS
running on the non-secure side. Single core v7-M targets will continue
to have PSA implemented via PSA emulation, implemented by Mbed OS.

Move or remove many PSA-implementing files, as PSA will be provided by
TF-M on non-single-v7-M targets. Delete any files that are not relevant
for PSA emulation mode.
 - Remove imported TF-M SPM
 - Remove Mbed SPM and tests
 - Remove Mbed-implemented PSA services and tests
 - Remove PSA_SRV_IMPL, PSA_SRV_IPC, PSA_SRV_EMUL and NSPE.
 - Replace PSA_SRV_EMUL and PSA_SRV_IMPL with MBED_PSA_SRV
 - Remove any files autogenerated by
   "tools/psa/generate_partition_code.py", which no longer exists.

Add new feature `PSA` to support PSA in Mbed OS.

Move the Mbed OS implementation of PSA services for v7-M targets (which
employ PSA emulation, and don't yet use TF-M) to
features/FEATURE_PSA/TARGET_MBED_PSA_SRV. Update the `requires`
attribute in TESTS/configs/baremetal.json to avoid breaking baremetal
testing builds.

Update .astyleignore to match new directory structure

Update Mbed TLS importer to place files into FEATURE_PSA

Create the following generic PSA targets:

* `PSA_Target` (Root level PSA generic target)
* `PSA_V7_M` (Single v7-M PSA generic target)
* `PSA_DUAL_CORE` (Dual-core PSA generic target)
* `PSA_V8_M` (v8-M PSA generic target)

Flatten MUSCA_NS and private MUSCA targets into public MUSCA targets.

Move mcuboot.bin to flat location (removing prebuilt folder)

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>
Signed-off-by: Jaeden Amero <jaeden.amero@arm.com>

urutva committed:
psa: Add a script to validate PSA targets

Add a script to parse `targets.json` to identify PSA targets and ensure
mandatory parameters are configured correctly for all PSA targets.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater committed:
cypress: psoc64: Add TF-M compatibility (again)

Partially revert f38e21fa6cf7 ("Update PSoC 6 BSPs to verion 1.2") to
restore TF-M compatibility.

Make the CY8CKIT_064S2_4343W target TF-M compatible by addding flash and
region definitions from TF-M (at c4f37c18c4a0) and by updating the
CY8CKIT_064S2_4343W linker script to create a flash image compatible
with TF-M.

Fixes: f38e21fa6cf7 ("Update PSoC 6 BSPs to verion 1.2")

Signed-off-by: Jaeden Amero <jaeden.amero@arm.com>

urutva committed:
psa: Import latest image signing scripts from TF-M

Import the latest image signing scripts from TF-M version
TF-Mv1.0-85-g3b7cc95a042c.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater committed:
psa: Add stubbed PSA client implementation

Add a stubbed PSA Client API implementation for application
compatibility. The interface is non-functional, but provides sane error
codes.

Patater committed:
Make PSA util compatible with Mbed Crypto 3.0.1

Mbed Crypto 3.0.1 ships with TF-M. To make Mbed TLS 2.22.0 compatible
with Mbed Crypto 3.0.1, changes are needed in psa_util.h (which
abstracts some portions of the PSA Crypto API for use with TLS) to deal
with new ECC curve define changes.

Signed-off-by: Jaeden Amero <jaeden.amero@arm.com>
Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater committed:
musca_b1: Remove device-specific virtual NVIC

Use instead the general TF-M v8-M virtual NVIC which will be added in
the commit that replaces Mbed PSA with TF-M PSA:
features/FEATURE_PSA/TARGET_TFM/TARGET_TFM_V8M/src/cmsis_nvic_virtual.c

jainvikas8 committed:
psa: Add TF-M utility scripts to python path

When python3 is enforced to build the ARM_MUSCA_A1 or ARM_MUSCA_B1
targets, it is unable to find binary utility tool scripts which are
imported from TF-M.

The reason to use the python3 environment is as follows: Mbed OS + TFM
contained a faulty boot record TLV, which failed the attestation test
(TF-M regression). The data in the boot record TLV will be included in
the generated attestation token as 1 item in the SW_COMPONENTS claim.

This data (in the boot record TLV) is pre-encoded in CBOR format at
build time and appended to the image during the image signing process
(done by the imgtool Python3 script).

Signed-off-by: Vikas Katariya <vikas.katariya@arm.com>

Patater committed:
psa: Fix style in psa_hrng.c

urutva committed:
tools: Improve PSA target identification logic

Current logic `is_TFM_target` relies on the availability of attribute
`tfm_target_name` to identify PSA targets. The API `is_TFM_target` is
used in pytest to validate PSA target configuration which again checks
the availability of `tfm_target_name`. If a target doesn't contain the
attribute `tfm_target_name` then this check will fail instead of
catching it. Therefore, we now check for `TFM` config option in `labels`
attribute.

The API `is_TFM_target()` returns true for Mbed OS PSA targets which are
supported by TF-M also.

Add a new API `is_PSA_target()` which returns true for all Mbed OS PSA
targets.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
tools: Musca B1 signing strategy

Currently, the final binary (TF-M + Mbed OS) is signed after
concatenating TF-M and Mbed OS binaries. But TF-M signs the images
separately and then concatenates them. Update the Musca B1 signing
strategy to match TF-M.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
target: Add BL2 macro to MUSCA B target config

BL2 macro is used in `region_defs.h` to define the `BL2_HEADER_SIZE`.
Without BL2 macro, `BL2_HEADER_SIZE` is set to 0. This leads to
incorrect start address (Reset_Handler of Mbed OS) derived by TF-M based
on `region_defs.h` and BL2 macro.

BL2 macro is set for MUSCA B in TF-M.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
target: Include missing `cmsis_nvic.h`

The header `cmsis_nvic.h` defines vector start address in RAM
`NVIC_RAM_VECTOR_ADDRESS` which is used in
`mbed_boot.c:mbed_cpy_nvic()`. But `mbed_boot.c` only includes
`cmsis.h`. Due to this `mbed_cpy_nvic` becomes an empty function and the
vectors don't get relocated to RAM. This causes BusFault error when Mbed
OS tries to update any of the IRQ handlers.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva commented on targets/targets.json:
```diff
@@ -858,7 +858,8 @@
             "KPSDK_CODE",
```
We should specify which targets will use `TARGET_MBED_PSA_SRV` in the commit message. I'd recommend `Make **single Armv7-M** PSA targets pick up files under TARGET_MBED_PSA_SRV.` and `Targets **(dual-core and Armv8-M)** that expect to use TF-M for their PSA implementation do not have MBED_PSA_SRV labels added to them.`

Patater replied:
Fixed in https://github.com/ARMmbed/mbed-os/pull/13003

urutva commented on features/mbedtls/importer/Makefile:
```diff
@@ -34,14 +34,16 @@ MBED_TLS_REPO_URL ?= git@github.com:ARMmbed/mbedtls.git
 TARGET_SRC:=$(TARGET_PREFIX)src
 TARGET_INC:=$(TARGET_PREFIX)inc
+TARGET_PSA:=$(TARGET_PREFIX)TARGET_MBED_PSA_SRV
+TARGET_PSA_INC:=$(TARGET_PREFIX)TARGET_MBED_PSA_SRV/inc
```
```suggestion
TARGET_PSA_INC:=$(TARGET_PSA)/inc
```

Patater replied:
Will fix after rebase on https://github.com/ARMmbed/mbed-os/pull/13003

jainvikas8 replied:
#13003 is merged, please fix this.

urutva commented on features/mbedtls/importer/Makefile:
```diff
@@ -34,14 +34,16 @@ MBED_TLS_REPO_URL ?= git@github.com:ARMmbed/mbedtls.git
+TARGET_SRV_IMPL:=$(TARGET_PSA)/COMPONENT_PSA_SRV_IMPL
 # COMPONENT_NSPE - Include code that compiles ONLY to the NSPE image and never
 # compiles to the SPE image
 TARGET_NSPE:=$(TARGET_SRV_IMPL)/COMPONENT_NSPE
```
We shouldn't have `COMPONENT_PSA_SRV_IMPL`  and `COMPONENT_NSPE` anymore.

Patater replied:
Will fix after rebase on https://github.com/ARMmbed/mbed-os/pull/13003

urutva commented on features/FEATURE_PSA/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,598 @@
+*Glossary*\
```
The commit message missing an `e` after `th`.

Patater replied:
O

urutva commented on .astyleignore:
```diff
@@ -1,11 +1,10 @@
 ^BUILD
```
The poring guide has been renamed to `psa_targets_in_mbed-os.md`. Please update the commit message `Add the document supporting_psa_in_mbed-os.md which describes how to add
PSA support for a target in Mbed 6.`

Also,
Create the following generic PSA targets:

* `PSA_Target` (Root level PSA target)
* `PSA_V7_M` (Single v7-M PSA generic target)
* `PSA_DUAL_CORE` (Dual-core PSA generic target)
* `PSA_V8_M` (v8-M generic PSA target)

Patater replied:
OK

urutva commented on LICENSE.md:
```diff
@@ -5,11 +5,11 @@ Folders containing files under different permissive license than Apache 2.0 are
 - [features/cryptocell/FEATURE_CRYPTOCELL310](./features/cryptocell/FEATURE_CRYPTOCELL310) - ARM Object Code and Header Files License
 - [features/FEATURE_BOOTLOADER](./features/FEATURE_BOOTLOADER) - PBL
 - [features/FEATURE_BLE/targets](./features/FEATURE_BLE/targets) - BSD-style, PBL, MIT-style
+- [features/FEATURE_PSA/FEATURE_TFM](./features/FEATURE_PSA/FEATURE_TFM) - BSD-3-Clause
```
Should we also list MCUBOOT and CBOR which are used by TF-M?

Patater replied:
Updated

urutva commented on features/FEATURE_PSA/TARGET_MBED_PSA_SRV/TESTS/crypto_init/main.cpp:
```diff
@@ -46,31 +46,6 @@ utest::v1::status_t greentea_test_setup(const size_t number_of_cases)
 
-static void check_multi_crypto_init_deinit()
-{
-#if !defined(COMPONENT_PSA_SRV_IPC)
```
This is a valid test case. The condition check `#if !defined(COMPONENT_PSA_SRV_IPC)` can be replaced with `#if !defined(FEATURE_PSA)`.

Patater replied:
Previously, if`COMPONENT_PSA_SRV_IPC` was not set, then this test would run only `TEST_SKIP()`. Now, `COMPONENT_PSA_SRV_IPC` is always unset, so this test, which would always skip, can be removed.

This PR doesn't aim to change the behavior of this test or extend the test to run in more places. It may be that we should have this test run on more platforms, but that's a separate concern.

urutva replied:
@Patater I believe `COMPONENT_PSA_SRV_IPC` flag was used to identify PSA targets. If it was not set then don't run this test. We can do similar with the flag `FEATURE_PSA`.
By removing this test case we are reducing the test coverage for PSA targets and as you said that's not the intention of this PR. So we should bring back this test case.

Patater replied:
`COMPONENT_PSA_SRV_IPC` was set for V7-dual NSPE, V7-dual SPE, V8-NS, and V8-S targets, never single v7-M. This test was always skipped on single v7-M. This test code is for the Mbed implementation of PSA. The Mbed implementation of PSA is no longer used for the targets that would have had `COMPONENT_PSA_SRV_IPC` set (dual core and v8-M).

urutva replied:
Ah that makes sense. Please go ahead and mark this conversation as resolved as Github doesn't allow marking conversations as resolved if one doesn't have write access to the repo.

urutva commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -24,8 +24,8 @@
 
-#if defined(TARGET_TFM) && defined(COMPONENT_NSPE)
-#include "TARGET_TFM/interface/include/tfm_ns_lock.h"
+#if defined(FEATURE_TFM) && defined(FEATURE_NSPE)
```
`FEATURE_NSPE` doesn't exist anymore.

Patater replied:
Fixed

urutva commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -64,9 +64,9 @@ MBED_NORETURN void mbed_rtos_start()
 #endif
 
-#if defined(TARGET_TFM) && defined(COMPONENT_NSPE)
+#if defined(FEATURE_TFM) && defined(FEATURE_NSPE)
```
`FEATURE_NSPE` doesn't exist anymore.

Patater replied:
Fixed

jainvikas8 commented on features/mbedtls/inc/mbedtls/psa_util.h:
```diff
@@ -163,6 +163,86 @@ static inline int mbedtls_psa_get_ecc_oid_from_id(
     char const **oid, size_t *oid_len )
 {
+#if TARGET_TFM
+    /* Use older Crypto API, at least until TF-M updates its crypto
```
Are we tracking this with an issue raised, probably might get forgotten?

Patater replied:
Filed https://github.com/ARMmbed/mbed-os/issues/13025. Will update comment to reference

jainvikas8 replied:
Please update the copyright year on the files for this commit.

jainvikas8 commented on features/mbedtls/src/pk.c:
```diff
@@ -617,12 +617,19 @@ int mbedtls_pk_wrap_as_opaque( mbedtls_pk_context *pk,
 
-    /* prepare the key attributes */
-    psa_set_key_type( &attributes, key_type );
     psa_set_key_bits( &attributes, bits );
```
Why `psa_set_key_bits( &attributes, bits );` is not set when the using `#if TARGET_TFM` ?
I see implementation in
`features/FEATURE_PSA/TARGET_TFM/include/psa/crypto.h`
`features/mbedtls/TARGET_MBED_PSA_SRV/inc/psa/crypto.h`

Patater replied:
Fixed

urutva committed:
tools: Remove obsolete API

The API is_PSA_non_secure_target() uses obsolete labels to detect if a
target is PSA non-secure target and is not needed anymore.

Mbed OS depends on TF-M for PSA SPM and services. TF-M is built using
it's own build system. Therefore, we don't need to differentiate secure
and non-secure targets anymore in Mbed OS as all PSA targets in Mbed OS
are non-secure targets.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

jainvikas8 commented on targets/TARGET_ARM_SSG/TARGET_MUSCA_B1/device/cmsis.h:
```diff
@@ -81,4 +81,8 @@
   #warning Not supported compiler type
 #endif
 
+#if defined(TARGET_MUSCA_B1)
```
Mbed OS than Mbed-OS in commit message?

Patater replied:
Fixed

jainvikas8 commented on features/FEATURE_PSA/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,598 @@
+## Building TF-M and running PSA compliance tests
+Follow the steps below to build TF-M and compliance tests:
+
+* Switch to `mbed-os-tf-m-regression-tests` directory
```
I think there is an instruction missing here - https://github.com/ARMmbed/mbed-os-tf-m-regression-tests/blob/master/main.cpp#L8 , we need to switch the example.

Patater replied:
That repo defaults to regression tests, which is what we want to run here, isn't it?

Ah, no this is for running the compliance tests, not regression tests.

Patater replied:
Fixed

jainvikas8 commented on features/mbedtls/importer/Makefile:
```diff
@@ -34,14 +34,16 @@ MBED_TLS_REPO_URL ?= git@github.com:ARMmbed/mbedtls.git
 TARGET_INC:=$(TARGET_PREFIX)inc
+TARGET_PSA:=$(TARGET_PREFIX)TARGET_MBED_PSA_SRV
+TARGET_PSA_INC:=$(TARGET_PREFIX)TARGET_MBED_PSA_SRV/inc
 TARGET_TESTS:=$(TARGET_PREFIX)TESTS
```
Please correct typo in commit message
"When TF-M `provies` the
PSA implementation on v8-M, we will add crypto_struct.h to the correct
place."

Patater replied:
Fixed in https://github.com/ARMmbed/mbed-os/pull/13003/ (next time @dgreen-arm pushes to it)

jainvikas8 reviewed (changes requested):
Please update copyright year to 2020, as needed for this PR.

jainvikas8 commented on features/FEATURE_PSA/psa_targets_in_mbed-os.md:
````diff
@@ -0,0 +1,600 @@
+*TIP*\
+Supported options:
+```console
+usage: build_tfm.py [-h]
````
It would be good to state `build_tfm.py -h` to see available options that avoids any updates to this section in the future if there any stales.

jainvikas8 commented on features/FEATURE_PSA/psa_targets_in_mbed-os.md:
````diff
@@ -0,0 +1,600 @@
+
+*TIP*\
+Supported options:
+```console
````
It would be good to state `build_psa_compliance.py -h` to see available options that avoid any updates to this section in the future if there any stales.

jainvikas8 commented on features/FEATURE_PSA/TARGET_TFM/TARGET_TFM_V8M/src/tfm_psa_ns_api.c:
```diff
@@ -1,19 +1,20 @@
 /*
- * Copyright (c) 2018, Arm Limited. All rights reserved.
+ * Copyright (c) 2018-2019, Arm Limited. All rights reserved.
```
If this file is not imported from TF-M, then copyright year needs to be changed. Also please add information in the commit message, that some files were imported from TF-M

Patater replied:
Yes, imported. I've updated the commit message to make this more clear. Note that the commit updates features/FEATURE_PSA/TARGET_TFM/VERSION.txt to indicate which version of TF-M the files came from.

jainvikas8 commented on features/FEATURE_PSA/TARGET_TFM/include/psa/client.h:
```diff
@@ -8,29 +8,56 @@
 #ifndef __PSA_CLIENT_H__
 #define __PSA_CLIENT_H__
 
+#include <stddef.h>
```
Please update copyright year in this commit for relevant files.

Patater replied:
Imported from TF-M

jainvikas8 commented on tools/targets/ARM_MUSCA_B1.py:
```diff
@@ -40,43 +40,63 @@ def musca_tfm_bin(t_self, non_secure_bin, secure_bin):
-    output.add_image(secure_bin, "SECURE")
-    output.add_image(non_secure_bin, "NON_SECURE")
+    #1. Run imgtool to sign the secure binary
+    sign_args = Namespace(
```
Are we sure that images are signed in the right way here?
https://git.trustedfirmware.org/trusted-firmware-m.git/tree/docs/user_guides/tfm_secure_boot.rst#n338 says as images are not concatenated and when signed separately

urutva replied:
The default value of `MCUBOOT_IMAGE_NUMBER` is 2 (https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/docs/user_guides/tfm_secure_boot.rst#n375). Therefore we are signing secure and non-secure separately and then concatenating them.

jainvikas8 replied:
> The default value of `MCUBOOT_IMAGE_NUMBER` is 2 (https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/docs/user_guides/tfm_secure_boot.rst#n375). Therefore we are signing secure and non-secure separately and then concatenating them.

Shouldn't we use - https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/docs/user_guides/tfm_secure_boot.rst#n342 to sign the non-secure side?

jainvikas8 commented on targets/TARGET_Cypress/TARGET_PSOC6/sb-tools/keys/USERAPP_CM4_KEY_PRIV.pem:
```diff
@@ -0,0 +1,5 @@
+-----BEGIN PRIVATE KEY-----
```
Are we adding a private key here?

Patater replied:
Yes

rwalton-arm commented on tools/psa/tfm/bin_utils/assemble.py:
```diff
@@ -25,7 +25,7 @@
 import os
 import shutil
-from . import macro_parser
+import macro_parser
```
This implicit relative import may cause issues in Python 3 depending on how this module is executed. We should keep the explicit relative import syntax.

Patater replied:
We need to make this change in TF-M.

Patater committed:
examples: Add PSA example to list of examples

jamesbeyond commented on tools/test/examples/examples.json:
```diff
@@ -333,6 +333,22 @@
+        "compile" : true,
+        "export": true,
+        "test" : true,
+        "baud_rate": 115200,
```
Is this has to be `115200`? can we use `9600`. This is a technical debt on Greentea, I but haven't got time to fix it

Patater replied:
OK, I'll update the example to use 9600 by default.

jamesbeyond reviewed (dismissed):
Oh, since in this PR you got so many file changes to worries about, I'll make my comment pronouncing, So you can't neglect that 😉

Patater committed:
psa: Configure Mbed TLS to use PSA as needed

Configure Mbed TLS to automatically enable PSA as needed. When Mbed OS
is configured to use PSA, configure Mbed TLS to use PSA. This prevents
leaking of the "how to make Mbed TLS use PSA" knowledge up into
targets.json, and thus makes porting simpler. There is now one place
where "how to make TLS use PSA" exists rather than repeated throughout
targets.json for each target that can't inherit from PSA_Target.
